### PR TITLE
[4.0] Shortcut Keys

### PR DESCRIPTION
**Pull request by brianteeman**
This is a proof of concept for a new plugin offering keyboard shortcuts in the Joomla admin based on previous work by @jneubauer

For testing I have created the following shortcut keys

Save - Alt/Opt S
Save & Close - Shift + Alt/Opt S
Save & New - Shift + Alt/Opt N
Save as Copy - Shift + Alt/Opt C
Cancel/Close - Alt/Opt X
New - Alt/Opt N
Help - Alt/Opt H
Options - Alt/Opt

Alt/Opt means Alt key on Windows/Linux and Option key on Mac

When testing please report the operating system, browser, language AND keyboard layout/lang

### CODE REVIEW

**Comment by dgrammatiko** [comment]
fwiw the `ctrl+s` was already implemented (but George manage to undo that) here: https://github.com/joomla/40-backend-template/pull/55

Also it worth copying the tinyMCE bits from there, so the combo works also when the cursor is in the editors iframe

**Comment by brianteeman** [comment]
As I said - by default NONE of these shortcuts work when the cursor is in a text area. Its just a config option to enable if desired

**Comment by dgrammatiko** [comment]
@brianteeman also you don't need an extra script, just add the logic in the toolbar.js

**Comment by brianteeman** [comment]
mousetrap is much better

**Comment by dgrammatiko** [comment]
> mousetrap is much better

I always prefer less code and of course no dependencies, but that's me...

**Comment by brianteeman** [comment]
I prefer maintainable code with no bus factor

**Comment by dgrammatiko** [comment]
> I prefer maintainable code with no bus factor

Just because you're pulling something from NPM is not automagically more maintainable or has a greater bus factor...

**Comment by ReLater** [comment]
Is a great and welcome feature!

**Comment by dgrammatiko** [comment]
Just to make my self clear here, when I said 
>  also you don't need an extra script, just add the logic in the toolbar.js

I meant the  `build/media_source/plg_system_shortcuts/js/shortcuts.js` not the mousetrap.js.
So all your code in the `shortcuts.js` can happily live in the toolbar, actually it'e exactly the place it belongs

**Comment by brianteeman** [comment]
Ah. I misunderstood you there.

As it's a proof of concept it's better on its own. If accepted it can be moved into that file of course

**Comment by brianteeman** [comment]
Also it depends if people want it to be configurable by the user (or even disabled) or not. And anyway toolbar is probably not the best place as even in this example there are some that work on pages without a toolbar

**Comment by dgrammatiko** [comment]
@brianteeman it can still be configurable, also by reading the code this works only with the toolbar buttons.
Also by reviewing the code of mousetrap.js I would say it's an overkill and probably not good fit for J4, reason the script is mainly polyfilling IE 6,7,8,9 which none is supported by J4

**Comment by zwiastunsw** [comment]
From the point of view of a11y, this is extremely important and necessary functionality. 
I miss the general toolbar shortcut key (of course, I remember that there will be quick access via SkipTo). My suggestion - instead of a shortcut Return to Control panel let the shortcut to the toolbar.
As to combination "j h" - I will ask my friend (blind accessibility expert) for their opinion.

**Comment by brianteeman** [comment]
The keys I used are just examples

**Comment by C-Lodder** [comment]
Great idea. Does this also include a hint list in the actual admin panel?

**Comment by brianteeman** [comment]
Not yet. But it's supported by mousetrap

**Comment by dgrammatiko** [comment]
@wilsonge fwiw if was still the JS leader I wouldn't accept this mousetrap script to enter the project's dependencies

File: build/build-modules-js/settings.json [hunk 1]
```diff
@@ -373,6 +373,12 @@
           "dist/metismenujs.min.js.map": "js/metismenujs.min.js.map"
         }
       },
+      "mousetrap": {
+        "name": "mousetrap",
+        "licenseFilename": "LICENSE",
+        "js": {
+          "mousetrap.js": "js/mousetrap.js",
+          "mousetrap.min.js": "js/mousetrap.min.js"
       "skipto": {
```
Comment (zero-24) [inline build/build-modules-js/settings.json]: it seams there is a syntax error in that file now?

Comment (brianteeman) [reply build/build-modules-js/settings.json]: oops

Comment (C-Lodder) [reply build/build-modules-js/settings.json]: add a `}` above this line

**Comment by brianteeman** [comment]
@zwiastunsw 
>  I will ask my friend (blind accessibility expert) for their opinion.

Did you get any feedback?

**Comment by zwiastunsw** [comment]
@brianteeman: I'm sory. Give me two days. I forgot.

**Comment by zwiastunsw** [comment]
This is the opinion of my blind friend (the accessibility tester):

The idea is OK and nothing should bite, but the semantically suggested shortcuts, at least in my opinion, are not associated with the actions they perform. 
Rather, I saw digitally-based shortcuts used to switch between page areas, bookmarks, etc. I've seen them. 
I don't know if any browser uses them to quickly select predefined bookmarks.
I would suggest shortcuts based on letters such as Ctrl+Shift+S - saves, Ctrl+Shift+X - output, etc. 
If necessary, I recommend to be inspired by Google documents or Wordpress.

(I checked - there are no shortcuts of this kind in Wordpress.)

**Comment by brianteeman** [comment]
If I remember correctly any shortcuts setup by the browser itself will override any shortcuts we set so we have to be careful about the keys being used

That's why I was testing if "j h" technically worked as a combination.

And I have no idea at all how this type of thing is handled in languages with non-latin keyboards

**Comment by wilsonge** [comment]
In this case I think we should not use the library you are proposing. the benefit in this specific case is minimal. 

For example the mod key is simply an alias for `/Mac|iPod|iPhone|iPad/.test(navigator.platform) ? 'meta' : 'ctrl'` (https://github.com/ccampbell/mousetrap/blob/master/mousetrap.js#L135) which isn't hard for us to port. And in *this* case I agree with @dgrammatiko that a library to effectively 1 modifier isn't much use for us. 

However I of course approve the concept. I can help with converting these to native JS if you would lik help. It really shouldn't be more than 30 minutes work.

**Comment by brianteeman** [comment]
I will put it on my todo list

**Comment by zwiastunsw** [comment]
My blind friend  has also drawn my attention to a very important issue. We must bear in mind the: [Success Criterion 2.1.4 Character Key Shortcuts](https://www.w3.org/TR/WCAG21/#character-key-shortcuts)

**Comment by wilsonge** [comment]
Because this is a plugin it can be disabled as any other plugin therefore we meet that criteria under "mechanism to disable". If we wanted to be really fancy in the future we could even investigate the ability to change the shortcuts as part of the plugin. But for now just disabling which we already have is the MVP to meet such compliance

Also none of these things Brian has done are only using `only letter (including upper- and lower-case letters), punctuation, number, or symbol characters.`. They all start with the modifier key (control on windows or the command key on mac)

So we're safe on multiple levels

**Comment by brianteeman** [comment]
Exactly!!

File: plugins/system/shortcuts/shortcuts.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+		{
+			$current_section = 0;
+		}
+
+		if (!($current_section & $section))
+		{
+			return false;
+		}
+
+		{
```
Comment (Quy) [inline plugins/system/shortcuts/shortcuts.php]: Delete?

File: plugins/system/shortcuts/shortcuts.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+		HTMLHelper::_('script', 'vendor/mousetrap/mousetrap.js', ['version' => 'auto', 'relative' => true], ['defer' => true]);
+		HTMLHelper::_('script', 'plg_system_shortcuts/shortcuts.js', array('version' => 'auto', 'relative' => true), ['defer' => true]);
+		}
```
Comment (Quy) [inline plugins/system/shortcuts/shortcuts.php]: Delete?

**Comment by brianteeman** [comment]
No need to comment on code style as this is NOT the code we will use. It
was just to test the concepts

**Comment by brianteeman** [comment]
This should now be in native js and not use any external libraries. I have updated the original description.

**Comment by brianteeman** [comment]
Sorry @dgrammatiko @wilsonge you were right - it didnt need the library at all

**Comment by brianteeman** [comment]
Updating branch to trigger drone

**Comment by wilsonge** [comment]
This works. However `altKey` either doesn't trigger or is overriden by browser behavior in chrome on my mac. option x still cuts, option s still downloads the web page, option h goes to the home page in chrome. all the commands are usable if using the actual `alt` key instead of the option key.

**Comment by brianteeman** [comment]
The docs say that alt on windows and opt on mac are triggered by `altKey ` https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/altKey

Can you go to https://keycode.info/ and tell me the results you get for alt and opt on your mac

**Comment by wilsonge** [comment]
Screenshot 2019-05-24 at 23 36 46
Screenshot 2019-05-24 at 23 36 52

This is alt and option

**Comment by brianteeman** [comment]
I can only think then that the mdn documentation I linked to is incorrect and they keystrokes are ALT on both and OSx

**Comment by wilsonge** [comment]
keyboard

Trying to figure it out tbh. Like the docs says the option key. that seems different to the command key. And the symbol is the one I have on my alt key.

I think the meta key is correct on macs https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/metaKey 

However saying that we start to override a bunch of browser short codes (nearly all the combinations you have are already native browser combinations)

**Comment by brianteeman** [comment]
Using the meta key really isn't an option

**Comment by brianteeman** [comment]
What needs to happen to move this forward?

It really needs some testers with non UK keyboards and non english language settings. @alikon @quy @SharkyKZ

**Comment by Quy** [comment]
Tested by placing cursor in Title field and pressing Alt + S. Got the same screenshot for Alt.

Windows 10, US standard keyboard layout.

Chrome 74.0.3729.169
Nothing happens.

Firefox 67.0.1. Dropdown menu appears.
shortcut

**Comment by brianteeman** [comment]
> Tested by placing cursor in Title field and pressing Alt + S. Got the same screenshot for Alt.

It is SHIFT and Alt

**Comment by Quy** [comment]
Same result.

**Comment by brianteeman** [comment]
might be a firefox bug - will investigate

**Comment by SharkyKZ** [comment]
In Edge, `Shift + Alt + S` result in Save, not Save & Close. `Alt + X` results in browser's `Settings and More` menu being opened. `Alt + O` doesn't do anything.

Windows 10, US layout.

**Comment by brianteeman** [comment]
I can confirm that the keyboard shortcuts I chose clash with some of the native browser shortcuts in firefox - grrh

**Comment by brianteeman** [comment]
I am closing this. It really serves little value especially regarding accessibility

**Pull request closed without merging**